subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -4628,28 +4534,129 @@ void lwm2m_socket_del(struct lwm2m_ctx *ctx)
 	}
 }
 
-/* LwM2M main work loop */
+static void create_notifications(sys_slist_t *observer_list,
+				 const int64_t timestamp)
+{
+	struct observe_node *obs;
+	int rc;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(observer_list, obs, node) {
+		if (obs->event_timestamp > obs->last_timestamp &&
+		    timestamp > obs->last_timestamp +
+		    MSEC_PER_SEC * obs->min_period_sec) {
+			rc = generate_notify_message(obs, true);
+			if (rc == -ENOMEM) {
+				/* no memory/messages available, retry later */
+				return;
+			}
+			obs->last_timestamp = timestamp;
+		} else if (timestamp >
+			   obs->last_timestamp +
+			   MSEC_PER_SEC * obs->max_period_sec) {
+			rc = generate_notify_message(obs, false);
+			if (rc == -ENOMEM) {
+				/* no memory/messages available, retry later */
+				return;
+			}
+			obs->last_timestamp = timestamp;
+		}
+	}
+}
 
-static void socket_receive_loop(void)
+static int socket_recv_message(struct lwm2m_ctx *client_ctx)
 {
 	static uint8_t in_buf[NET_IPV6_MTU];
-	static struct sockaddr from_addr;
 	socklen_t from_addr_len;
 	ssize_t len;
-	int i;
+	static struct sockaddr from_addr;
+
+	from_addr_len = sizeof(from_addr);
+	len = recvfrom(client_ctx->sock_fd, in_buf, sizeof(in_buf) - 1,
+		       MSG_DONTWAIT, &from_addr, &from_addr_len);
```

Comment:
nit: The socket was already put in the `O_NONBLOCK` state, `MSG_DONTWAIT` is redundant here.

Reply:
Good catch, was a leftover from earlier modifications.